Make -auth option work better

Add -auth option handling for Xming (patch3)
Use an internally generated cookie for authentication of server clients when running the -auth option in -multiwindow mode. (patch11)

XXX: doesn't internal clipboard client need similar authentication?
XXX: does this need some thinking about other security mechanisms we have now...


Copyright (C) Colin Harrison 2005-2008
http://www.straightrunning.com/XmingNotes/
http://sourceforge.net/projects/xming/ 

---
 xserver/hw/xwin/InitOutput.c       |    4 +++-
 xserver/hw/xwin/winglobals.c       |    3 +++
 xserver/hw/xwin/winmultiwindowwm.c |   13 +++++++++++++
 xserver/hw/xwin/winprocarg.c       |   25 +++++++++++++++++++++++++
 4 files changed, 44 insertions(+), 1 deletion(-)


Index: xorg-server-1.5.2/xserver/hw/xwin/winmultiwindowwm.c
===================================================================
--- xorg-server-1.5.2.orig/xserver/hw/xwin/winmultiwindowwm.c
+++ xorg-server-1.5.2/xserver/hw/xwin/winmultiwindowwm.c
@@ -81,6 +81,7 @@ extern void winDebug(const char *format,
 #endif
 #define WIN_JMP_OKAY		0
 #define WIN_JMP_ERROR_IO	2
+#define AUTH_NAME		"MIT-MAGIC-COOKIE-1"
 
 
 /*
@@ -130,6 +131,10 @@ typedef struct _XMsgProcArgRec {
 
 extern char *display;
 extern void ErrorF (const char* /*f*/, ...);
+#if defined(XCSECURITY)
+extern unsigned int	g_uiAuthDataLen;
+extern char		*g_pAuthData;
+#endif
 
 
 /*
@@ -1237,6 +1242,14 @@ winInitMultiWindowWM (WMInfoPtr pWMInfo,
 
   /* Print the display connection string */
   ErrorF ("winInitMultiWindowWM - DISPLAY=%s\n", pszDisplay);
+
+#if defined(XCSECURITY)
+  /* Use our generated cookie for authentication */
+  XSetAuthorization (AUTH_NAME,
+		     strlen (AUTH_NAME),
+		     g_pAuthData,
+		     g_uiAuthDataLen);
+#endif
   
   /* Open the X display */
   do
Index: xorg-server-1.5.2/xserver/hw/xwin/winprocarg.c
===================================================================
--- xorg-server-1.5.2.orig/xserver/hw/xwin/winprocarg.c
+++ xorg-server-1.5.2/xserver/hw/xwin/winprocarg.c
@@ -1,6 +1,7 @@
 /*
 
 Copyright 1993, 1998  The Open Group
+Copyright (C) Colin Harrison 2005-2008
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
@@ -57,6 +58,7 @@ extern char *			g_pszLogFile;
 extern Bool			g_fLogFileChanged;
 #endif
 extern Bool			g_fXdmcpEnabled;
+extern Bool			g_fAuthEnabled;
 extern char *			g_pszCommandLine;
 extern Bool			g_fKeyboardHookLL;
 extern Bool			g_fNoHelpMessageBox;                     
@@ -1289,6 +1291,29 @@ ddxProcessArgument (int argc, char *argv
     }
 
   /*
+   * Look for the '-auth' argument
+   */
+  if (IS_OPTION ("-auth"))
+    {
+#ifdef  __MINGW32__
+      HANDLE hFile;
+      char * pszFile;
+      CHECK_ARGS (1);
+      pszFile = argv[++i];
+      hFile = CreateFile(pszFile,GENERIC_READ,FILE_SHARE_READ,NULL,OPEN_EXISTING,FILE_ATTRIBUTE_NORMAL,NULL);
+      if (hFile == INVALID_HANDLE_VALUE)
+	winMessageBoxF ("This authorization file for the -auth option could not be opened...\n"
+			"\"%s\"\n"
+			"You should use an \"Xauthority\" file in your HOME directory.\n"
+			"\nIgnoring and continuing.\n",
+			MB_ICONINFORMATION,
+			pszFile);
+#endif
+      g_fAuthEnabled = TRUE;
+      return 0; /* Let DIX parse this again */
+    }
+
+  /*
    * Look for the '-indirect' or '-broadcast' arguments
    */
   if (IS_OPTION ("-indirect")
Index: xorg-server-1.5.2/xserver/hw/xwin/InitOutput.c
===================================================================
--- xorg-server-1.5.2.orig/xserver/hw/xwin/InitOutput.c
+++ xorg-server-1.5.2/xserver/hw/xwin/InitOutput.c
@@ -1,6 +1,7 @@
 /*
 
 Copyright 1993, 1998  The Open Group
+Copyright (C) Colin Harrison 2005-2008
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
@@ -73,6 +74,7 @@ extern int			g_iLogVerbose;
 Bool				g_fLogInited;
 
 extern Bool			g_fXdmcpEnabled;
+extern Bool			g_fAuthEnabled;
 #ifdef HAS_DEVWINDOWS
 extern int			g_fdMessageQueue;
 #endif
@@ -1031,7 +1033,7 @@ InitOutput (ScreenInfo *screenInfo, int 
 
 #if defined(XCSECURITY)
   /* Generate a cookie used by internal clients for authorization */
-  if (g_fXdmcpEnabled)
+  if (g_fXdmcpEnabled || g_fAuthEnabled)
     winGenerateAuthorization ();
 #endif
 
Index: xorg-server-1.5.2/xserver/hw/xwin/winglobals.c
===================================================================
--- xorg-server-1.5.2.orig/xserver/hw/xwin/winglobals.c
+++ xorg-server-1.5.2/xserver/hw/xwin/winglobals.c
@@ -1,5 +1,6 @@
 /*
  *Copyright (C) 2003-2004 Harold L Hunt II All Rights Reserved.
+ *Copyright (C) Colin Harrison 2005-2008
  *
  *Permission is hereby granted, free of charge, to any person obtaining
  * a copy of this software and associated documentation files (the
@@ -26,6 +27,7 @@
  *from Harold L Hunt II.
  *
  * Authors:	Harold L Hunt II
+ *              Colin Harrison
  */
 
 #ifdef HAVE_XWIN_CONFIG_H
@@ -58,6 +60,7 @@ HWND		g_hDlgExit = NULL;
 HWND		g_hDlgAbout = NULL;
 const char *	g_pszQueryHost = NULL;
 Bool		g_fXdmcpEnabled = FALSE;
+Bool           g_fAuthEnabled = FALSE;
 HICON		g_hIconX = NULL;
 HICON		g_hSmallIconX = NULL;
 #ifndef RELOCATE_PROJECTROOT
